File: pypi/trackfw/serve/api_metrics.py
"""
serve/api_metrics.py — Metrics API: lead time, cycle time, burndown.
Reutiliza _parse_log de commands/metrics.py — sem reimplementação.
Espelho Python de internal/serve/api_metrics.go e npm/src/serve/api_metrics.js.
"""

import os
from datetime import datetime, timedelta

from trackfw.commands.metrics import _parse_log

STATES = ["wip", "backlog", "blocked", "done", "abandoned"]


def _state_distribution(cfg):
    """
    Conta arquivos .md por estado varrendo os diretórios de roadmap.
    """
    roadmap_dir = cfg.get("roadmap_dir", "docs/roadmaps")
    namespacing = cfg.get("roadmap_namespacing", "flat")
    distribution = {s: 0 for s in STATES}

    def _count_dir(dir_path, state):
        if not os.path.isdir(dir_path):
            return 0
        try:
            return sum(
                1 for f in os.listdir(dir_path)
                if f.endswith(".md") and not os.path.isdir(os.path.join(dir_path, f))
            )
        except OSError:
            return 0

    if namespacing == "by_agent":
        agents = cfg.get("agents") or []
        if not agents:
            try:
                agents = [
                    e for e in os.listdir(roadmap_dir)
                    if os.path.isdir(os.path.join(roadmap_dir, e))
                ]
            except OSError:
                agents = []
        for agent in agents:
            for state in STATES:
                distribution[state] += _count_dir(os.path.join(roadmap_dir, agent, state), state)
    else:
        for state in STATES:
            distribution[state] += _count_dir(os.path.join(roadmap_dir, state), state)

    return distribution


def _calc_lead_time(transitions):
    """
    Lead time: do primeiro movimento (qualquer → backlog) até done.
    Retorna média em dias.
    """
    by_name = {}
    for tr in transitions:
        by_name.setdefault(tr["basename"], []).append(tr)

    lead_times = []
    for entries in by_name.values():
        start_ts = None
        done_ts = None
        for e in sorted(entries, key=lambda x: x["timestamp"]):
            if start_ts is None:
                start_ts = e["timestamp"]
            if e["to"] == "done":
                done_ts = e["timestamp"]
        if start_ts is not None and done_ts is not None:
            delta_days = (done_ts - start_ts).total_seconds() / 86400.0
            lead_times.append(delta_days)

    if not lead_times:
        return 0.0
    return sum(lead_times) / len(lead_times)


def _calc_cycle_time(transitions):
    """
    Cycle time: da entrada em wip até done.
    Retorna média em dias.
    """
    by_name = {}
    for tr in transitions:
        by_name.setdefault(tr["basename"], []).append(tr)

    cycle_times = []
    for entries in by_name.values():
        wip_ts = None
        done_ts = None
        for e in sorted(entries, key=lambda x: x["timestamp"]):
            if e["to"] == "wip" and wip_ts is None:
                wip_ts = e["timestamp"]
            if e["to"] == "done":
                done_ts = e["timestamp"]
        if wip_ts is not None and done_ts is not None:
            delta_days = (done_ts - wip_ts).total_seconds() / 86400.0
            cycle_times.append(delta_days)

    if not cycle_times:
        return 0.0
    return sum(cycle_times) / len(cycle_times)


def _calc_abandonment_rate(transitions):
    """
    Taxa de abandono: count(abandoned) / count(abandoned + done).
    """
    by_name = {}
    for tr in transitions:
        by_name.setdefault(tr["basename"], []).append(tr)

    done_count = 0
    abandoned_count = 0
    for entries in by_name.values():
        final_states = {e["to"] for e in entries}
        if "abandoned" in final_states:
            abandoned_count += 1
        elif "done" in final_states:
            done_count += 1

    total = done_count + abandoned_count
    if total == 0:
        return 0.0
    return abandoned_count / total


def _calc_burndown(transitions):
    """
    Burndown semanal: para cada semana, quantos roadmaps estão abertos e fechados.
    Retorna lista de { date, open, closed }.
    """
    if not transitions:
        return []

    timestamps = [tr["timestamp"] for tr in transitions]
    min_ts = min(timestamps)
    max_ts = max(timestamps)

    # Normalizar para início da semana (segunda-feira)
    start = min_ts - timedelta(days=min_ts.weekday())
    start = start.replace(hour=0, minute=0, second=0, microsecond=0)
    end = max_ts + timedelta(days=7)

    # Construir eventos por basename
    by_name = {}
    for tr in transitions:
        by_name.setdefault(tr["basename"], []).append(tr)

    burndown = []
    current = start
    while current <= end:
        week_end = current + timedelta(days=7)
        open_count = 0
        closed_count = 0

        for entries in by_name.values():
            # Pegar o estado final até o fim da semana
            state_at_week = None
            for e in sorted(entries, key=lambda x: x["timestamp"]):
                if e["timestamp"] <= week_end:
                    state_at_week = e["to"]
            if state_at_week is None:
                pass
            elif state_at_week in ("done", "abandoned"):
                closed_count += 1
            else:
                open_count += 1

        burndown.append({
            "date": current.strftime("%Y-%m-%d"),
            "open": open_count,
            "closed": closed_count,
        })
        current = week_end

    return burndown


def get_metrics(cfg):
    """
    Calcula e retorna métricas de entrega a partir do .trackfw-log.
    """
    log_path = os.path.join(cfg.get("roadmap_dir", "docs/roadmaps"), ".trackfw-log")
    transitions = _parse_log(log_path)

    lead_time = _calc_lead_time(transitions)
    cycle_time = _calc_cycle_time(transitions)
    abandonment_rate = _calc_abandonment_rate(transitions)
    state_distribution = _state_distribution(cfg)
    burndown = _calc_burndown(transitions)

    return {
        "lead_time_avg_days": round(lead_time, 2),
        "cycle_time_avg_days": round(cycle_time, 2),
        "abandonment_rate": round(abandonment_rate, 4),
        "state_distribution": state_distribution,
        "burndown": burndown,
    }

File: npm/src/serve/api_metrics.js
'use strict'

const fs = require('fs')
const path = require('path')
const { parseLog, calculate } = require('../commands/metrics')

const STATES = ['wip', 'backlog', 'blocked', 'done', 'abandoned']
const MS_PER_DAY = 24 * 60 * 60 * 1000

/**
 * countFilesInState conta arquivos .md em um diretório (sem recursão).
 * @param {string} dir
 * @returns {number}
 */
function countFilesInState(dir) {
  try {
    return fs.readdirSync(dir).filter(f => {
      if (!f.endsWith('.md')) return false
      try { return !fs.statSync(path.join(dir, f)).isDirectory() } catch (_) { return false }
    }).length
  } catch (_) {
    return 0
  }
}

/**
 * buildStateDistribution conta roadmaps por estado.
 * @param {object} cfg
 * @returns {Object<string, number>}
 */
function buildStateDistribution(cfg) {
  const dist = { wip: 0, backlog: 0, blocked: 0, done: 0, abandoned: 0 }
  const roadmapDir = cfg.roadmapDir || 'docs/roadmaps'
  const namespacing = cfg.roadmapNamespacing || 'flat'

  if (namespacing === 'by_agent') {
    let agents = cfg.agents || []
    if (!agents.length) {
      try {
        agents = fs.readdirSync(roadmapDir).filter(f => {
          try { return fs.statSync(path.join(roadmapDir, f)).isDirectory() } catch (_) { return false }
        })
      } catch (_) { agents = [] }
    }
    for (const agent of agents) {
      for (const state of STATES) {
        dist[state] = (dist[state] || 0) + countFilesInState(path.join(roadmapDir, agent, state))
      }
    }
  } else {
    for (const state of STATES) {
      dist[state] = (dist[state] || 0) + countFilesInState(path.join(roadmapDir, state))
    }
  }

  return dist
}

/**
 * buildBurndown calcula pontos semanais a partir do log de transições.
 * @param {Array} transitions
 * @returns {Array<{date: string, open: number, closed: number}>}
 */
function buildBurndown(transitions) {
  if (!transitions.length) return []

  let minTs = Infinity
  let maxTs = -Infinity
  for (const tr of transitions) {
    const ms = tr.timestamp.getTime()
    if (ms < minTs) minTs = ms
    if (ms > maxTs) maxTs = ms
  }

  // Agrupar por semana (domingo como início)
  const weekMap = new Map()

  function weekKey(ms) {
    const d = new Date(ms)
    // arredondar para o domingo anterior
    const day = d.getDay() // 0=domingo
    const sunday = new Date(ms - day * MS_PER_DAY)
    return sunday.toISOString().slice(0, 10)
  }

  for (const tr of transitions) {
    const wk = weekKey(tr.timestamp.getTime())
    if (!weekMap.has(wk)) weekMap.set(wk, { open: 0, closed: 0 })
    const entry = weekMap.get(wk)
    if (tr.to === 'done' || tr.to === 'abandoned') {
      entry.closed++
    } else if (tr.to === 'wip' || tr.to === 'backlog') {
      entry.open++
    }
  }

  const sorted = Array.from(weekMap.entries()).sort(([a], [b]) => a.localeCompare(b))
  return sorted.map(([date, v]) => ({ date, open: v.open, closed: v.closed }))
}

/**
 * handleMetrics responde ao GET /api/metrics com JSON de métricas.
 * @param {object} cfg
 * @param {http.IncomingMessage} req
 * @param {http.ServerResponse} res
 */
function handleMetrics(cfg, req, res) {
  const roadmapDir = cfg.roadmapDir || 'docs/roadmaps'
  const logPath = path.join(roadmapDir, '.trackfw-log')

  const transitions = parseLog(logPath)
  const metrics = calculate(transitions)

  // Converter cycle time de ms para dias
  const cycleTimeAvgDays = metrics.cycleTimeMeanMs > 0
    ? parseFloat((metrics.cycleTimeMeanMs / MS_PER_DAY).toFixed(2))
    : 0

  // Lead time: da criação (primeira entrada em qualquer estado) até done
  // Usando as mesmas transições — aproximar como cycle time se não houver campo de criação
  // (o campo creation_date não está no log; usamos start → done como lead time também)
  const leadTimeAvgDays = cycleTimeAvgDays

  // Abandonment rate: abandonados / (done + abandonados)
  const dist = buildStateDistribution(cfg)
  const totalCompleted = (dist.done || 0) + (dist.abandoned || 0)
  const abandonmentRate = totalCompleted > 0
    ? parseFloat(((dist.abandoned || 0) / totalCompleted).toFixed(4))
    : 0

  const burndown = buildBurndown(transitions)

  const result = {
    lead_time_avg_days: leadTimeAvgDays,
    cycle_time_avg_days: cycleTimeAvgDays,
    abandonment_rate: abandonmentRate,
    state_distribution: dist,
    burndown,
  }

  res.writeHead(200, { 'Content-Type': 'application/json' })
  res.end(JSON.stringify(result))
}

module.exports = { handleMetrics }
